cmd/f3/manifest.go

```diff
@@ -47,6 +55,155 @@ var manifestGenCmd = cli.Command{
+			var err error
+			id, _, err = crypto.GenerateEd25519Key(rand.Reader)
+			if err != nil {
+				return xerrors.Errorf("generating random libp2p identity: %w", err)
+			}
+		}
+
+		peerID, err := peer.IDFromPrivateKey(id)
+		if err != nil {
+			return xerrors.Errorf("getting peer ID from libp2p identity: %w", err)
+		}
+		_, _ = fmt.Fprintf(c.App.Writer, "Manifest server peer ID: %s\n", peerID)
+
+		host, err := libp2p.New(
+			libp2p.Identity(id),
+			libp2p.ListenAddrStrings(c.StringSlice("listenAddr")...),
+			libp2p.UserAgent("f3-dynamic-manifest-server"),
+		)
+		if err != nil {
+			return xerrors.Errorf("initializing libp2p host: %w", err)
+		}
+
+		// Connect to all bootstrap addresses once. This should be sufficient to build
+		// the pubsub mesh, if not then we need to periodically re-connect and/or pull in
+		// the Lotus bootstrapping, which includes DHT connectivity.
+		//
+		// For now, simply connect to bootstrappers and test if it suffices before doing
+		// extra work.
+		for _, bootstrapper := range c.StringSlice("bootstrapAddr") {
+			addr, err := peer.AddrInfoFromString(bootstrapper)
+			if err != nil {
+				return xerrors.Errorf("parsing bootstrap address %s: %w", bootstrapper, err)
+			}
+			if err := host.Connect(c.Context, *addr); err != nil {
+				return xerrors.Errorf("connecting to bootstrapper %s: %w", bootstrapper, err)
+			}
+		}
+
+		manifestPath := c.String("manifest")
+		loadManifestAndVersion := func() (manifest.Manifest, manifest.Version, error) {
+
+			m, err := loadManifest(manifestPath)
+			if err != nil {
+				return manifest.Manifest{}, "", xerrors.Errorf("loading manifest: %w", err)
+			}
+			version, err := m.Version()
+			if err != nil {
+				return manifest.Manifest{}, "", xerrors.Errorf("versioning manifest: %w", err)
+			}
+			return m, version, nil
+		}
+
+		initManifest, manifestVersion, err := loadManifestAndVersion()
+		if err != nil {
+			return xerrors.Errorf("loading initial manifest: %w", err)
+		}
+
+		pubSub, err := pubsub.NewGossipSub(c.Context, host, pubsub.WithPeerExchange(true))
+		if err != nil {
+			return xerrors.Errorf("initialzing pubsub: %w", err)
+		}
+
+		sender, err := manifest.NewManifestSender(host, pubSub, initManifest, c.Duration("publishInterval"))
+		if err != nil {
+			return xerrors.Errorf("initialzing manifest sender: %w", err)
+		}
+		_, _ = fmt.Fprintf(c.App.Writer, "Started manifest sender with version: %s\n", manifestVersion)
+
+		go func() {
+			sender.Start(c.Context)
+		}()
+		defer func() {
+			sender.Stop()
+			_ = host.Close()
+		}()
+
+		checkTicker := time.NewTicker(c.Duration("checkInterval"))
+		for c.Context.Err() == nil {
+			select {
+			case <-c.Context.Done():
+				return c.Context.Err()
+			case <-checkTicker.C:
+				if nextManifest, nextManifestVersion, err := loadManifestAndVersion(); err != nil {
+					_, _ = fmt.Fprintf(c.App.ErrWriter, "Failed reload manifest: %v\n", err)
+				} else if manifestVersion != nextManifestVersion {
```

Comment:
Do we want to force manifest updates to be strictly increasing by 1?

Reply:
discussed in persion: varsion is hex encoded hash of content